### DISCUSSION TITLE
Cannot get footer or header to work

### DISCUSSION THREAD

**Discussion by PericlesTheo** [gotenberg/gotenberg#722] [Q&A]
👋 Hey there,

I have a basic example and I'm trying to add headers and footers to it. However, even though when rendering the HTML itself works, when combining them into the PDF it fails. ~**I'm using Gotenberg 6.2**~. UPDATE: I'm using 7.9, I mistakenly thought it was 6.2.

[example_v1.pdf](https://github.com/gotenberg/gotenberg/files/13416825/example_v1.pdf)

You really need to zoom in to see the header and footer. The body on the other hand renders perfectly.

<details>
  <summary><bold>You can see the request body here</bold></summary>
  
 ```
 --423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="paperHeight"

11
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="marginBottom"

0.5
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="marginRight"

0.5
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="marginLeft"

0.5
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="preferCssPageSize"

false
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="printBackground"

false
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="omitBackground"

false
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="landscape"

false
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="nativePageRanges"


--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="marginTop"

0.5
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="pdfFormat"


--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="paperWidth"

8.5
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="scale"

1.0
--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="index.html"; filename="index.html"
Content-Type: application/octet-stream

<html>
    <head>
        <meta charset="UTF-8" />
        <title>Example PDF</title>
        <style type="text/css">html, body {
  color: #1d1d1b;
}</style>
    </head>
    <body>
        <h1>Body</h1>
        <p>Account holder: ABC THERE</p>
        <br/>
        <p>Account 1234</p>
        <h1>Partial</h1>
        <p>This is my partial</p>
        <p>Account holder: ABC THERE</p>
        <br/>
        <p>Account 1234</p>    </body>
</html>

--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="header.html"; filename="header.html"
Content-Type: application/octet-stream

<html>
    <head>
        <meta charset="UTF-8" />
        <style type="text/css">html, body {
  color: #1d1d1b;
}</style>
    </head>
    <body>
        <p>This is my header</p>
    </body>
</html>

--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e
Content-Disposition: form-data; name="footer.html"; filename="footer.html"
Content-Type: application/octet-stream

<html>
    <head>
        <meta charset="UTF-8" />
        <style type="text/css">html, body {
  color: #1d1d1b;
}</style>
    </head>
    <body>
        <p>This is my footer</p>
    </body>
</html>

--423db45531196d92c57daf27c0851f0739e9af8325395f16022e1cec844e--
  
 ```
</details>

I've tried increasing [the font size as well](https://gotenberg.dev/docs/routes#header--footer) but it was the same issue. Another issue I spotted is that the footer was not respected and the body would bleed into the footer.

Let me know if you need anything else for me to supply.

**Comment by gulien** [comment]
Hello @PericlesTheo,

Gotenberg 6 is no more maintained 😅 

Please consider upgrading to Gotenberg 7, and reopen the issue if you still struggle to make it work.

Also, the `6.x` documentation may be found here: https://gotenberg.dev/docs/6.x/introduction

**Comment by PericlesTheo** [comment]
@gulien not so fast 😆 I just realised we are actually using 7.9.

I was looking at the wrong place.

**Comment by gulien** [comment]
Using [the following test data](https://github.com/gotenberg/gotenberg/tree/main/test/testdata/chromium/html), this is was I got (using the demo instance):

[exemple_demo.pdf](https://github.com/gotenberg/gotenberg/files/13418223/exemple_demo.pdf)

**Comment by PericlesTheo** [comment]
Thanks!

So I just applied the font-sizes based on the test data provided above and now I can see the header and footer but other problems arise, specifically the body and header bleed into each other.

<details>
  <summary><bold>You can see the request body here</bold></summary>

```
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="marginLeft"

0.5
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="scale"

1.0
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="nativePageRanges"


--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="pdfFormat"


--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="marginBottom"

0.5
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="paperHeight"

11
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="marginTop"

0.5
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="marginRight"

0.5
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="preferCssPageSize"

false
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="paperWidth"

8.5
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="omitBackground"

false
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="landscape"

false
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="printBackground"

false
--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="index.html"; filename="index.html"
Content-Type: application/octet-stream

<html>
    <head>
        <meta charset="UTF-8" />
        <title>Example PDF</title>
        <style type="text/css">html, body {
  color: #1d1d1b;
}</style>
    </head>
    <body>
        <h1>Body</h1>
        <p>Account holder: ABC THERE</p>
        <br/>
        <p>Account 1234</p>
        <h1>Partial</h1>
        <p>This is my partial</p>
        <p>Account holder: ABC THERE</p>
        <br/>
        <p>Account 1234</p>    </body>
</html>

--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="header.html"; filename="header.html"
Content-Type: application/octet-stream

<html>
    <head>
        <meta charset="UTF-8" />
        <style type="text/css">html, body {
  color: #1d1d1b;
}</style>
        <style>
            body {
                font-size: 8rem;
                margin: 4rem auto;
            }
        </style>
    </head>
    <body>
        <p>This is my header</p>
    </body>
</html>

--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70
Content-Disposition: form-data; name="footer.html"; filename="footer.html"
Content-Type: application/octet-stream

<html>
    <head>
        <meta charset="UTF-8" />
        <style type="text/css">html, body {
  color: #1d1d1b;
}</style>
        <style>
            body {
                font-size: 30rem;
                margin: 4rem auto;
            }
        </style>
    </head>
    <body>
        <p>This is my footer</p>
    </body>
</html>

--36317e372c403246d47ff5fb95e9403fafb172e754d7ff1aa3f620f1cc70--
```

</details>

[example_v2.pdf](https://github.com/gotenberg/gotenberg/files/13423473/example_v2.pdf)

I think the questions I have overall are:
- How come the header and footer bleed into each other in my case?
- How come we need to increase the font-size for the footer and header so much in order to be visible compared to the body?
- The body has a specific margin around the page, however the header and footer start from the edge. If one wanted to align the two, how could we go about it?

**Reply by krwi** [reply]
I had the same problem the other day.
Try adding margins to your `index.html` style like so:
```
        @page {
            margin-top: 50mm;
            margin-bottom: 40mm;
        }
```
In other words, you basically tell your main document (`index.html`) to respect some margins at the top and bottom, as it doesn't know about any headers or footers that will be applied later on (and their height).

**Comment by gulien** [comment]
> How come the header and footer bleed into each other in my case?

I guess you need more margins (1?).

> How come we need to increase the font-size for the footer and header so much in order to be visible compared to the body?

No idea, it's just how it works with Chromium 😅 

> The body has a specific margin around the page, however the header and footer start from the edge. If one wanted to align the two, how could we go about it?

There is no magic trick here; you'll have to play with your margins / font sizes to find the sweet spot.

Maybe a workaround would be to put your header / footer directly inside the body. If you're rendering your own HTML, it's doable I guess.

**Reply by PericlesTheo** [reply]
> Maybe a workaround would be to put your header / footer directly inside the body

Maybe it would work for the header but unfortunately I don't think it works with the footer. Imagine you have a long text. Unless you explicitly specify when the next page starts, you will not be able to show the footer on every page.

I will do a bit more investigation and see but there aren't any settings I have missed to control this right?

**Reply by gulien** [reply]
You might be right about the footer, but there is maybe a trick with CSS print properties.

> I will do a bit more investigation and see but there aren't any settings I have missed to control this right?

Everything I know is available in the documentation 😄 https://gotenberg.dev/docs/routes#header--footer
